disk/test/com/treode/disk/edit/DriveGroupSpec.scala

```diff
@@ -82,8 +87,100 @@ class DriveGroupSpec extends FlatSpec with DiskChecks {
     override def toString = "DriveGroupPhase"
   }
 
+  /** Convenient methods for testing. */
+  implicit class RichDiskController (controller: DiskController) {
+
+    // TODO: Move this into the DiskController trait.
+    def attach (attaches: DriveAttachment*): Async [Notification] =
+      controller.asInstanceOf [DiskAgent] .change (DriveChange (attaches, Seq.empty))
+
+    // TODO: Move this into the DiskController trait.
+    def attach (geom: DriveGeometry, paths: Path*): Async [Notification] =
+      attach (paths map (DriveAttachment (_, geom)): _*)
+
+    // NOT TODO: This is for testing only.
+    def attach (paths: String*) (implicit geom: DriveGeometry): Async [Notification] =
+      attach (geom, paths map (Paths.get (_)): _*)
+
+    def drain (paths: Path*): Async [Notification] =
+      controller.asInstanceOf [DiskAgent] .change (DriveChange (Seq.empty, paths))
+
+    def drain (paths: String*) (implicit geom: DriveGeometry): Async [Notification] =
+      drain (paths map (Paths.get (_)): _*)
+  }
+
+  implicit val config = DiskTestConfig()
+  implicit val events = new StubDiskEvents
+  implicit val geom = DriveGeometry (8, 6, 1 << 14)
+
+  "DriveGroup.change" should "reject nonexistant files" in {
+    implicit val files = new StubFileSystem
+    implicit val scheduler = StubScheduler.random()
+    implicit val recovery = new RecoveryAgent
+    val launch = recovery.reattach().expectPass()
+    launch.launch()
+    val controller = launch.controller
+    val e = controller.attach ("a") .expectFail [IllegalArgumentException]
+    assertResult ("requirement failed: File a does not exist.") (e.getMessage)
+  }
+
+  it should "reject duplicate filenames (at the same time)" in {
+    implicit val files = new StubFileSystem
+    implicit val scheduler = StubScheduler.random()
+    implicit val recovery = new RecoveryAgent
+    files.create(Seq ("f1", "f2", "f3") map (Paths.get (_)), 0, 1 << 14)
+    val launch = recovery.reattach().expectPass()
+    launch.launch()
+    val controller = launch.controller
```

Comment:
These methods are doing the same setup, and that they are doing so up to `controller`. I think you should extract that common code into a method. You could make a method like:

```
def setup (paths: String*): Controller = ...
```